### ISSUE TITLE
Create api-availability measurement

### ISSUE THREAD

**Issue by mm4tt** [kubernetes/perf-tests#1096]
### Justification

See this comment - https://github.com/kubernetes/perf-tests/issues/1086#issuecomment-593918574

#### V0

1. Create a new "ApiAvailability" measurement
1. This measurement should periodically probe the apiserver's /healthz endpoint and record whether api was available or not
1. The measurement should output the following stats
    1. availability percentage (e.g. % of OK responses over all responses)
    1. Longest consecutively unavailability period
    1. ...

#### V1
1. Make the measurement fail if apiserver is consecutively  not available in the last XX min (e.g. 30min). 
1. Make the error critical to make sure the test execution is stopped in such case

#### V2

1. Come up with exact SLI/SLO definition, make it available in perf-dash for further analysis
1. Make it a WIP [Scalability SLO](https://github.com/kubernetes/community/blob/master/sig-scalability/slos/slos.md)
1. Evaluate and promote to official SLO

/good-first-issue

**Comment by vamossagar12** [comment]
/assign
I would like to take this up. Would these be more details furnished?

**Comment by mm4tt** [comment]
Great, thanks @vamossagar12 

Let's start with the V0. Could you let me know which part of the description is not clear or requires further details?

**Comment by vamossagar12** [comment]
thanks @mm4tt Couple of things:
1) how frequently should the probe happen? Would it be configurable or we can define it to be fixed?
2) The availability percentage should be wrt the entire perf test run and we have to keep track of percentage and longest unavailable period?

As an outside question, typically if the health end point is down for a configured time, response would be to scale up or something like that. Just wanted to know the rationale behind adding this particular measurement. Thanks!

**Comment by wojtek-t** [comment]
> how frequently should the probe happen? Would it be configurable or we can define it to be fixed?

Seems like implementation detail. We can make it configurable, but I doubt we will be using different values once we agree on sth.

> The availability percentage should be wrt the entire perf test run and we have to keep track of percentage and longest unavailable period?

Correct

> As an outside question, typically if the health end point is down for a configured time, response would be to scale up or something like that. Just wanted to know the rationale behind adding this particular measurement. Thanks!

There are couple points:
- we want to ensure that we understand how availability looks like (I can imgine all other SLOs being fine, but cluster being periodicall unavailable)
- we can use it as optimization - if cluster is down for X minutes, the it probably won't get up ever, and we can shut down the test and save money

**Comment by vamossagar12** [comment]
Thanks for the updates @wojtek-t . I will start with this issue now. Pretty sure there would be a few more questions along the way though :)

**Comment by vamossagar12** [comment]
hi. Since I last commented, I didn't get a chance to look at it. Will start over the weekend..

**Comment by vamossagar12** [comment]
hi so i started looking at this today. One of the things is that there's already a measurement metrics_for_e2e which, among other things, also fetches the metrics from api-server. 

So, the new measurement that needs to be created, can work along the same lines just that instead of hitting the /metrics end-point, it can hit the /healthz end-point of api-server. Thats' from an implementation standpoint.

The other question that I had is that the measurements that we defines are within a Step and a group of Steps forms a single test, so when we say that this new measurement measures the health of the api-server for the duration of a test(mentioned above), so what exactly does a _test_ mean in this case? A Step which houses the measurement(s) or the overall test?

**Comment by mm4tt** [comment]
Hey, @vamossagar12 

It makes sense to make the implementation similar to the metrics_for_e2e. It's exactly as you said, instead of hitting /metrics, we'll be hitting /healthz.

Regarding your second question. Usually a measurement has two actions: `start` and `gather`. This one should work the same. The `start` action should start a goroutine that will be continuously pinging /healthz endpoint and the `gather` will stop this goroutine and wrap everything up. Does it make sense?

**Comment by wojtek-t** [comment]
> It makes sense to make the implementation similar to the metrics_for_e2e. It's exactly as you said, instead of hitting /metrics, we'll be hitting /healthz.

Not sure I understand - for metrics_for_e2e IIRC we're fetching the metrics once at the end of test. Assuming that the above is true - it can't really work the same way...

**Comment by mm4tt** [comment]
> > It makes sense to make the implementation similar to the metrics_for_e2e. It's exactly as you said, instead of hitting /metrics, we'll be hitting /healthz.
> 
> Not sure I understand - for metrics_for_e2e IIRC we're fetching the metrics once at the end of test. Assuming that the above is true - it can't really work the same way...

Good point, I should have checked how the metrics_for_e2e works :) Still you should be able to take some inspiration from that measurement. Let me know if you have more questions.

**Comment by vamossagar12** [comment]
Actually I meant to use the metrics_e2e as only a baseline of how to interact with the api server. I hadn't looked at the internals so what @wojtek-t told is even more valuable :)

**Comment by vamossagar12** [comment]
> Hey, @vamossagar12
> 
> It makes sense to make the implementation similar to the metrics_for_e2e. It's exactly as you said, instead of hitting /metrics, we'll be hitting /healthz.
> 
> Regarding your second question. Usually a measurement has two actions: `start` and `gather`. This one should work the same. The `start` action should start a goroutine that will be continuously pinging /healthz endpoint and the `gather` will stop this goroutine and wrap everything up. Does it make sense?

Regarding point 2, I still have a question. As far as what I have understood, the hierarchy of a test is as follows:

Test -> Step -> Phase(s) Or measurement(s)

If we just focus on measurements, it's 2 levels lower than a test. So, when we hit the api-server from start and gather from within a measurement, that would be still within the context of that particular step.  Is the question clear now? Or is my thinking totally off track here 😄

**Comment by mm4tt** [comment]
Oh, I see where the confusion comes from. The hierarchy you listed is correct. But the measurement there should be treated as measurement invocation. The measurement object life spans the whole test. So if your test is

```
Step1: call method start on measurement A
Step2: do something else
Step3: do something else
Step4: Call method gather on measurement A
```

Then methods start and gather will be called on the same measurement instance

**Comment by vamossagar12** [comment]
@mm4tt I have taken an initial stab at this here: https://github.com/kubernetes/perf-tests/pull/1162/files

Plz review when you have the bandwidth..

**Comment by vamossagar12** [comment]
@wojtek-t The PR got merged. I guess the next task would be to write the config files? Any specific areas for me to start looking at for that?

**Comment by wojtek-t** [comment]
/remove-lifecycle stale

We roughly have milestone V0 implemented at this point.

**Comment by wojtek-t** [comment]
/remove-lifecycle stale

/assign @tosi3k 

@tosi3k - can you please briefly describe where we are with it now?

**Comment by tosi3k** [comment]
We're in the middle of V1. We have already added a configurable availability percentage threshold under which the test would fail.

The problem is that the API availability measurement makes the API call latency measurement fail. This is because the former runs `kubectl exec` periodically underneath - this results in POST pods' exec subresource API call latency jump above the 1s threshold (which is understandable and shouldn't be considered as an API call latency SLO violation). @jupblb is currently working on fixing this.
